examples/aligned_accessor/aligned_accessor.cpp

```diff
@@ -0,0 +1,714 @@
+// https://github.com/rapidsai/raft/pull/725#discussion_r937991701
+
+namespace {
+
+using test_value_type = float;
+constexpr int min_overalignment_factor = 8;
+constexpr int min_byte_alignment =
+  static_cast<int>(min_overalignment_factor * sizeof(float));
+
+// Use int, not size_t, as the index_type.
+// Some compilers have trouble optimizing loops with unsigned or 64-bit index types.
+using index_type = int;
+
+namespace stdex = std::experimental;
+
+// Prefer std::assume_aligned if available, as it is in the C++ Standard.
+// Otherwise, use a compiler-specific equivalent if available.
+
+// NOTE (mfh 2022/08/08) BYTE_ALIGNMENT must be unsigned and a power of 2.
+#if defined(__cpp_lib_assume_aligned)
+#  define _MDSPAN_ASSUME_ALIGNED( ELEMENT_TYPE, POINTER, BYTE_ALIGNMENT ) (std::assume_aligned< BYTE_ALIGNMENT >( POINTER ))
+  constexpr char assume_aligned_method[] = "std::assume_aligned";
+#elif defined(__ICL)
+#  define _MDSPAN_ASSUME_ALIGNED( ELEMENT_TYPE, POINTER, BYTE_ALIGNMENT ) POINTER
+  constexpr char assume_aligned_method[] = "(none)";
+#elif defined(__ICC)
+#  define _MDSPAN_ASSUME_ALIGNED( ELEMENT_TYPE, POINTER, BYTE_ALIGNMENT ) POINTER
+  constexpr char assume_aligned_method[] = "(none)";
+#elif defined(__clang__)
+#  define _MDSPAN_ASSUME_ALIGNED( ELEMENT_TYPE, POINTER, BYTE_ALIGNMENT ) POINTER
+  constexpr char assume_aligned_method[] = "(none)";
+#elif defined(__GNUC__)
+  // __builtin_assume_aligned returns void*
+#  define _MDSPAN_ASSUME_ALIGNED( ELEMENT_TYPE, POINTER, BYTE_ALIGNMENT ) reinterpret_cast< ELEMENT_TYPE* >(__builtin_assume_aligned( POINTER, BYTE_ALIGNMENT ))
+  constexpr char assume_aligned_method[] = "__builtin_assume_aligned";
+#else
+#  define _MDSPAN_ASSUME_ALIGNED( ELEMENT_TYPE, POINTER, BYTE_ALIGNMENT ) POINTER
+  constexpr char assume_aligned_method[] = "(none)";
+#endif
+
+// Some compilers other than Clang or GCC like to define __clang__ or __GNUC__.
+// Thus, we order the tests from most to least specific.
+#if defined(__ICL)
+#  define _MDSPAN_ALIGN_VALUE_ATTRIBUTE( BYTE_ALIGNMENT ) __declspec(align_value( BYTE_ALIGNMENT ));
+  constexpr char align_attribute_method[] = "__declspec(align_value(BYTE_ALIGNMENT))";
+#elif defined(__ICC)
+#  define _MDSPAN_ALIGN_VALUE_ATTRIBUTE( BYTE_ALIGNMENT ) __attribute__((align_value( BYTE_ALIGNMENT )));
+  constexpr char align_attribute_method[] = "__attribute__((align_value(BYTE_ALIGNMENT)))";
+#elif defined(__clang__)
+#  define _MDSPAN_ALIGN_VALUE_ATTRIBUTE( BYTE_ALIGNMENT ) __attribute__((align_value( BYTE_ALIGNMENT )));
+  constexpr char align_attribute_method[] = "__attribute__((align_value(BYTE_ALIGNMENT)))";
+#else
+#  define _MDSPAN_ALIGN_VALUE_ATTRIBUTE( BYTE_ALIGNMENT )
+  constexpr char align_attribute_method[] = "(none)";
+#endif
+
+template<class ElementType>
+constexpr bool
+valid_byte_alignment(const std::size_t byte_alignment,
+		     const std::size_t size_of_element)
+{
+  return byte_alignment >= alignof(ElementType); // && is a power of two
```

Comment:
compare to `bit_floor(byte_alignment)` when C++20 is available?

Reply:
@dalg24 Excellent suggestion; thank you!  I should be able to use the feature test macro `__cpp_lib_int_pow2` for these C++20 functions.

Reply:
Fixed; thank you!